custom/m4change/filters.py

```diff
@@ -33,3 +39,72 @@ def options(self):
             ("delivery", _("delivery")),
             ("immunization", _("immunization"))
         ]
+
+
+class RestrictedLocationDrillDown(object):
+
+    def __init__(self, domain, user):
+        self.domain = domain
+        self.user = user
+
+    def _get_path_to_first_common_parent(self):
+        sql_locations = self.user.get_sql_locations(self.domain)
+        list_of_ancestors = [
+            sql_location.get_ancestors(ascending=True, include_self=True)
+            for sql_location in sql_locations
+        ]
+
+        if list_of_ancestors:
+            common_ancestors = set(list_of_ancestors[0])
+            for ancestors in list_of_ancestors[1:]:
+                common_ancestors = common_ancestors.intersection(set(ancestors))
+
+            for ancestors in list_of_ancestors:
+                for ancestor in ancestors:
+                    yield ancestor
+                    if ancestor in common_ancestors:
+                        break
+
+    def get_locations_json(self):
+        def loc_to_json(loc):
+            return {
+                'pk': loc.pk,
+                'name': loc.name,
+                'location_type': loc.location_type.name,
+                'uuid': loc.location_id,
+                'is_archived': loc.is_archived,
+                'level': loc.level,
+                'parent_id': loc.parent_id
+            }
+        user = self.user
+
+        user_locations = list(SQLLocation.objects.accessible_to_user(self.domain, user))
+        if not user_locations:
+            return []
+
+        user_locations.extend(list(self._get_path_to_first_common_parent()))
+        user_locations = list(set(user_locations))
```

Comment:
`SQLLocation.objects.get_queryset_ancestors` would come in useful here.  I believe you can just do:
```python
user_locations = SQLLocation.objects.get_queryset_ancestors(
    SQLLocation.objects.accessible_to_user(self.domain, user),
    include_self=True
)
```
And that would give all accessible locations plus their ancestry.